server/src/main/java/org/apache/cassandra/sidecar/livemigration/LiveMigrationFileDownloader.java

```diff
@@ -0,0 +1,688 @@
+
+        if (toDownloadPercentage <= successThreshold)
+        {
+            return Future.succeededFuture(filesToDownload);
+        }
+
+        LOGGER.info("{} Download size met the successThreshold({}).{}",
+                    logPrefix,
+                    successThreshold,
+                    " Skipping download of " + filesToDownload.size() + " files of size " + downloadSize);
+        return Future.succeededFuture(Collections.emptyList());
+    }
+
+    @VisibleForTesting
+    Future<OperationStatus> sortBySizeAndDownload(List<InstanceFileInfo> instanceFileInfoList)
+    {
+        List<Future<Void>> futureList = new ArrayList<>(instanceFileInfoList.size());
+
+        // Sorting by size in descending order so that large files will be downloaded first. This will implicitly put
+        // directories at the end (because size is explicitly set to -1 for directories).
+        instanceFileInfoList.sort(Collections.reverseOrder(Comparator.comparingLong(instanceFileInfo -> instanceFileInfo.size)));
+
+        List<Callable<Future<Void>>> downloadTasks = new ArrayList<>();
+        SidecarInstanceImpl instance = new SidecarInstanceImpl(source, port);
+
+
+        for (InstanceFileInfo file : instanceFileInfoList)
+        {
+            if (file.fileType.equals(FileType.DIRECTORY))
+            {
+                futureList.add(createDirectory(file));
+            }
+            else if (file.fileType.equals(FileType.FILE) && file.size == 0)
+            {
+                futureList.add(createEmptyFile(file));
+            }
+            else
+            {
+                Callable<Future<Void>> task = () -> getDownloadTask(instance, file);
+                downloadTasks.add(task);
+            }
+        }
+
+        // start the file downloads with a cap on concurrency
+        concurrentTaskExecutor = new AsyncConcurrentTaskExecutor<>(vertx, downloadTasks, request.maxConcurrency);
+        List<Future<Void>> taskFutures = concurrentTaskExecutor.start();
+        futureList.addAll(taskFutures);
+
+        return Future.join(futureList)
+                     .compose(ar -> {
+                                  if (ar.succeeded())
+                                  {
+                                      return Future.succeededFuture(this.updateState(OperationStatus::toDownloadCompleteState));
+                                  }
+                                  LOGGER.warn("{} Failed to download few files. Updating state to {}", logPrefix, State.FAILED);
+                                  return Future.succeededFuture(this.updateState(OperationStatus::tryFailureState));
+                              },
+                              ar -> {
+                                  LOGGER.error("{} Failed to download files", logPrefix, ar);
+                                  return Future.succeededFuture(this.updateState(OperationStatus::tryFailureState));
+                              });
```

Comment:
Handle failures in one place and simplify a bit

```suggestion
        return Future.join(futureList)
                     .compose(res -> res.succeeded() ? Future.succeededFuture() : Future.failedFuture(res.cause()))
                     .compose(ar -> Future.succeededFuture(this.updateState(OperationStatus::toDownloadCompleteState)),
                              cause -> {
                                  LOGGER.error("{} Failed to download files.  Updating state to {}", logPrefix, State.FAILED, cause);
                                  return Future.succeededFuture(this.updateState(OperationStatus::tryFailureState));
                              });
```

Reply:
done